lineage/BaumWelch.py

```diff
@@ -145,6 +150,31 @@ def do_M_E_step(tHMMobj, gammas):
         tHMMobj.estimate.E[state_j].estimator(all_cells, all_gammas[:, state_j])
 
 
+def do_M_E_step_atonce(all_tHMMobj, all_gammas):
+    """ perform the M_E step when all the concentrations are given at once. """
+    all_cells = []
+    gms = []
+    # gather all gammas from different concentrations and append.
+    for i, tHMMobj in enumerate(all_tHMMobj):
+        all_gms = np.vstack(all_gammas[i])
+        gms.append(all_gms)
+        all_cells.append([cell.obs for lineage in tHMMobj.X for cell in lineage.output_lineage])
+    # reshape the gammas so that each list in this list of lists is for each state.
+    final_gm = []
+    for j in range(all_tHMMobj[0].num_states):
+        tmp1 = []
+        for aray in gms:
+            tmp1.append(aray[:, j])
+        final_gm.append(tmp1)
+
+    # for each state, estimate the parameters and assign them to the tHMMobj
+    for state_j in range(all_tHMMobj[0].num_states):
+        output = atonce_estimator(all_cells, final_gm[state_j])
```

Comment:
I'd check that you're collecting all of the data together, and then assigning the results, correctly before worrying about the estimator. You can do this by simply skipping the actual estimator call and assigning the same parameter values as were there before.

Reply:
Should I collect all the cells and all the gammas when I am running the functions on each state? Or should I only pass those cells and those gammas for each specific state? I thought I should pass all and the gammas are there for distinguishing the states... I'm not sure I'm right

Reply:
Functions should only receive the data they need for the action at hand.